[ETCM-355] Send fork id to peers

lukasz-golebiewski opened the pull request:
# Description
First part of https://eips.ethereum.org/EIPS/eip-2364. Extends the ETH64 protocol Status by a fork id field and broadcasts it to peers during handshake 
 
Does not yet contain peer fork id validation

lukasz-golebiewski committed:
[ETCM-355] Introduce ETH64 message format

lukasz-golebiewski committed:
[ETCM-355] Send fork id in the Status message

lukasz-golebiewski committed:
[ETCM-355] Remove capabilities from EtcHandshakerConfiguration

dzajkowski commented on src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala:
```diff
@@ -27,13 +27,15 @@ case class EtcHelloExchangeState(handshakerConfiguration: EtcHandshakerConfigura
-    Capability.negotiate(hello.capabilities.toList, handshakerConfiguration.capabilities) match {
+    Capability.negotiate(hello.capabilities.toList, handshakerConfiguration.blockchainConfig.capabilities) match {
       case Some(ProtocolVersions.ETC64) => EtcNodeStatus64ExchangeState(handshakerConfiguration)
       case Some(ProtocolVersions.ETH63) => EtcNodeStatus63ExchangeState(handshakerConfiguration)
```
request: do you mind changing this to Eth? I missed it and it's confusing this way.

dzajkowski commented:
overall looks good to me. question: would you be able to deploy to staging to see how it behaves? I guess running it locally against ETC mainnet would also be a good idea.

lukasz-golebiewski committed:
[ETCM-355] Rename EtcNodeStatus63ExchangeState -> EthNodeStatus63ExchangeState

leo-bogastry reviewed (approved):
I thinks it's looking good

lukasz-golebiewski commented:
Waiting before merging for https://github.com/input-output-hk/mantis/pull/1029

lukasz-golebiewski committed:
[ETCM-355] Use eth/64 in the nomad test env

lukasz-golebiewski committed:
fixup! [ETCM-355] Send fork id in the Status message

lukasz-golebiewski committed:
[ETCM-355] scalafix

lukasz-golebiewski committed:
[ETCM-355] Log protocol name upon unknown message type

lukasz-golebiewski committed:
[ETCM-355] Log negotiated protocol version

lukasz-golebiewski committed:
[ETCM-355] Log received messages in PeerActor

lukasz-golebiewski committed:
[ETCM-355] Post merge fix

lukasz-golebiewski committed:
[ETCM-355] Expand MessageDecoderSpec with eth/64 tests

lukasz-golebiewski committed:
[ETCM-355] Introduce ProtocolFamily

This also fixes a bug in NewBlock message creation, where only
protocol version was taken into account

dzajkowski commented on src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala:
```diff
@@ -45,10 +45,12 @@ class BlockBroadcast(val etcPeerManager: ActorRef) {
+      val remoteStatus = peers(peer.id).peerInfo.remoteStatus
+
+      val message: MessageSerializable = remoteStatus.protocolFamily match {
+        case ETH => blockToBroadcast.as63
```
would be so cool to have this fail if a new block type is introduced. just a thought, no great ideas.

lukasz-golebiewski replied:
Not sure what you mean here

dzajkowski replied:
thinking out loud:
```scala
  private def broadcastNewBlock(blockToBroadcast: BlockToBroadcast, peers: Map[PeerId, PeerWithInfo]): Unit =
    obtainRandomPeerSubset(peers.values.map(_.peer).toSet).foreach { peer =>
      val remoteStatus = peers(peer.id).peerInfo.remoteStatus

      val message: MessageSerializable = {
        val capOpt = Capabilities.of(remoteStatus.protocolFamily, remoteStatus.protocolVersion)
        capOpt match {
          case Capabilities.Eth63Capability => blockToBroadcast.as63
          case Capabilities.Eth64Capability => blockToBroadcast.as63
          case Capabilities.Etc64Capability => blockToBroadcast.as64
          case None                         => throw new RuntimeException("Unknown capability, cannot broadcast block")
        }
      }

...

    def of(pf: ProtocolFamily, v: Int): Option[Capability] =
      (pf, v) match {
        case (ProtocolFamily.ETH, 63) => Some(Eth63Capability)
        case (ProtocolFamily.ETH, 64) => Some(Eth64Capability)
        case (ProtocolFamily.ETC, 64) => Some(Etc64Capability)
        case _                        => None
      }
```

and an ADT out of the capabilities?

dzajkowski commented on src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala:
```diff
@@ -31,18 +31,15 @@ case class EtcHelloExchangeState(handshakerConfiguration: EtcHandshakerConfigura
-      case Some(ProtocolVersions.ETH64) => {
+      case Some(ProtocolVersions.ETH64) =>
         log.debug("Negotiated protocol version with client {} is eth/64", hello.clientId)
         EthNodeStatus64ExchangeState(handshakerConfiguration)
```
could you also fix the 'orElse' clause that I introduced (`orElse(WireDecoder)`)

lukasz-golebiewski replied:
I would like to do this in a follow up PR, this one has become quite old and heavy
